### PR TITLE
Add ~ and ^ semver semantic to nimble "requires"

### PR DESCRIPTION
**Pull request by marcomq**
Added ~ and ^ semver semantic to nimble "requires" 
Check https://devhints.io/semver for semver  examples
#907

### CODE REVIEW

File: src/nimblepkg/version.nim [hunk 1]
```diff
@@ -316,12 +366,33 @@ when isMainModule:
...
   doAssert findLatest(parseVersionRange(">= 0.1 & <= 0.4"), versions) ==
       (newVersion("0.2.3"), "v0.2.3")
+  doAssert findLatest(parseVersionRange("^= 0.1"), versions) ==
+      (newVersion("0.1.1"), "v0.1.1")
+  doAssert findLatest(parseVersionRange("^= 0.0.1"), versions) ==
+      (newVersion("0.0.1"), "v0.0.1")
+  doAssert findLatest(parseVersionRange("^= 2.2.2"), versions) ==
+      (newVersion("2.3.2"), "v2.3.2")
+  doAssert findLatest(parseVersionRange("^2.2.2"), versions) ==
+      (newVersion("2.3.2"), "v2.3.2")
```
Comment (dom96) [inline src/nimblepkg/version.nim:391]: Hmm, nit but I think I'd rather enforce the need to write `^=`

File: src/nimblepkg/version.nim [hunk 2]
```diff
@@ -153,6 +153,46 @@ proc makeRange*(version: string, op: string): VersionRange =
...
   result.ver = Version(version)
 
+proc parseVersionStringAt(s: string, i: var int): string = 
+  ## parse version string after ~ or ^ and return it
+  while i < s.len:
+    case s[i]
+    of '0'..'9', '.':
+      result.add(s[i])
+    of 'x', 'X', '*': # only allowed for ~ and ^
+      result.add("0")
```
Comment (dom96) [inline src/nimblepkg/version.nim:163]:
```suggestion
    of '*': # only allowed for ~ and ^
      result.add("0")
```

Let's only support one variant, much easier to learn then.

File: src/nimblepkg/version.nim [hunk 3]
```diff
@@ -153,6 +153,46 @@ proc makeRange*(version: string, op: string): VersionRange =
...
+    if numbers[n] == "0":
+      inc(zeros)
+    else: break
+  let distanceToMajor = if majorCompatibility: 0 else: 1
+  var excludedPosition = min(numbers.len - 2, zeros + distanceToMajor)
+  if zeros == 2: 
+    excludedPosition = max(2, excludedPosition)
+  numbers[excludedPosition] = $(numbers[excludedPosition].parseInt() + 1)
+  var zeroPosition = excludedPosition + 1
+  while(zeroPosition < numbers.len):
```
Comment (dom96) [inline src/nimblepkg/version.nim:191]:
```suggestion
  while zeroPosition < numbers.len:
```

**Review by dom96 — APPROVED** [review]
Awesome, thank you!

Approving, but I'd like to see some docs for this, can you write some explaining what these new operators mean exactly in the readme? :)

**Comment by marcomq** [comment]
Thx for the review :) I will update the readme. 
After checking the semver documentation again, I have to check if everything is still correct for 2 or 4 digit version numbers. Maybe I also need to add some test cases.

**Comment by marcomq** [comment]
Sry - after creating some documentation, I made some fixes and refactored a lot of code. However, I think the solution is better than my previous one. Let me know if I should make some additional changes.

**Review by dom96 — APPROVED** [review]
Looks good, some suggestions for follow up:

* Create tests that verify the version range created for these new intersections, I think that is a much clearer way to test these :)


Might be possible to simplify `getNextIncompatibleVersion`, but let's get it merged, we can iterate later. As long as we have tests it's all good :)

**Pull request merged**